XpressAI/xircuits
# ✨Add branch component

# Description

This will add a branch component to enable branching of the workflow. This also indirectly bypass the `checkAllNodesConnected` when there's at least a connection from `Start` to `Finish`. Last node from both `If` port can be any node except **Finish node**. **Finish node** must only be connected from `Finished ▶` port. `Finished ▶` port of branch component **MUST** be connected. 

This also fix `hyperparameter boolean` doesn't give correct value.

Branch component have one `inPort` and two `outPorts`:

**`inPort`**

1. condition (`bool`) : Can be left empty. Default = True. 

**`outPorts`**

1. `If True  ▶` : Can be left empty. If no condition is set and is connected, this will run. 
2. `If False ▶` : Can be left empty. 
3. `Finished ▶` : This will always run. **Must** connect with a node.

![Branch with finished Workflow](https://user-images.githubusercontent.com/84708008/170155470-089a1e00-84cc-4b7e-8b42-1b3b58336e4e.gif)

## Pull Request Type

- [x] Others (Python code generator)

## Type of Change


# Tests

1. Try create a workflow using a branch component.
   1. Try using literal node to set the `condition`.
2. Try creating multiple workflow using multiple branch component.
   1. Try using hyperparameter node to set the `condition`.
3. (Additional) Xircuits can run when there's at least have one complete workflow(Start to Finish).
   1. Try creating a complete workflow.
   2. Add another component. Don't connect the node.
   3. Try running it. It's expected to work.
4. Try compiling workflow without ending with Finish node.
   1. An error tooltip will be display. 
5. Try running without connecting one/both `If` port but `Finished ▶` port **must** be connected.
   1. If either/both port not connected, it will run the `Finished ▶` port.
6. Try running without connecting `Finished ▶` port.
   1. An error tooltip will be display. 
7. Try different and crazy but feasible workflow. (Let me know if it doesn't work)

## Tested on?


# Notes

It's inside `Custom` Folder. Also, not sure why but in some area(right side) on canvas, the branch node get distorted. 
There can be only one **Finish** node.
